design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,142 @@
+
+## Integration plan with Rook
+
+The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There should be option in `operator.yaml` to validate whether the COSI controller exists and another option at object store level to bring up the Ceph COSI driver. Each Ceph object store should have its own COSI driver.
+
+### Pre-requisites
+
+- COSI CRDs should be installed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
+```
+
+- COSI controller should be deployed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
+```
+
+- Create Service Account and RBAC for COSI driver
+
+```bash
+kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml
+kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml
+```
+
+### End to End Workflow
+
+#### Changes in Operator.yaml
+
+Below option validates whether the COSI controller is deployed in the cluster. If it is not deployed, it will log an warning message and when user tries to create object store with cosi, it will reconcile until the controller is deployed.
+
+```yaml
+  env:
+  - name: COSI_CONTROLLER_ENABLED
+    value: "true"
+```
+
+#### Changes in CephObjectStore CRD
+
+Following option will bring up the Ceph COSI driver for the RGW server.
+
+```yaml
+  cephCosiDrvierSpec:
+    enabled: true
+```
+
+By setting this option, Rook will deploy Ceph COSI driver with sidecar after creating the Ceph object store and will also create `CephObjectStoreUser` objectstorage-provisioner-<object-store-name> which provides credentials for Ceph COSI driver. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster which already present with current codebase. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for multiple ceph object stores in the cluster. Rook also need to support update of COSI related images independently.
+
+#### Updating existing CephObjectStore Deployment
+
+If the user wants to enable the COSI driver for existing ceph object store, the user need to update the `cephCosiDrvierSpec` option in the CephObjectStore CRD. Rook will bring up the COSI driver for the existing ceph object store.
```

Comment:
No need for this if we just detect the cosi controller automatically

Reply:
I also think that there's no need to detect if the COSI controller is running. If the object store just always creates the secret/configmap resources that the ceph-cosi controller will look for to understand when a CephObjectSgtore is present, then COSI will be able to use said object store for creating buckets without anything new needing to be added to the API.